applications/pipeline_visualization/python/pipeline_visualization.py

```diff
@@ -0,0 +1,269 @@
+from holoscan.conditions import PeriodicCondition
+from holoscan.core import Application, Operator, OperatorSpec
+
+from holohub.nats_logger import NatsLogger
+
+
+class SourceOp(Operator):
+    """Source operator that generates a sine wave signal.
+
+    This operator produces a synthetic time-series signal consisting of a sine wave
+    with a gradually increasing frequency (10-20 Hz). It outputs 3000 samples per compute cycle.
+    """
+
+    def __init__(self, fragment, *args, **kwargs):
+        super().__init__(fragment, *args, **kwargs)
+        self.frequency = 10.0  # Current frequency of the sine wave in Hz
+
+    def setup(self, spec: OperatorSpec):
+        spec.output("out")
+
+    def compute(self, op_input, op_output, context):
+        """Generate a sine wave signal with time-varying frequency."""
+        # Generate a sine wave signal with time-varying frequency
+        samples = 3000  # Number of samples in the signal
+        duration = 1.0  # Duration of signal in seconds
+        sample_time = duration / samples
+        omega = 2 * np.pi * self.frequency  # Angular frequency
+
+        # Generate the sine wave
+        t = np.arange(samples) * sample_time
+        wave = np.sin(omega * t)
+
+        # Reshape to (samples, 1) to match C++ implementation
+        wave = wave.reshape(samples, 1).astype(np.float32)
+
+        # Gradually increase frequency from 10 to 20 Hz, then wrap back to 10 Hz
+        self.frequency += 0.1
+        if self.frequency > 20.0:
+            self.frequency = 10.0
+
+        # Emit the tensor to the output port
+        op_output.emit(dict(wave=as_tensor(wave)), "out")
+
+
+class ModulateOp(Operator):
+    """Modulation operator that adds high-frequency noise to the input signal.
+
+    This operator receives a time-series signal and adds a 300 Hz sinusoidal modulation
+    with small amplitude (0.05) to simulate measurement noise or signal perturbation.
+    """
+
+    def setup(self, spec: OperatorSpec):
+        spec.input("in")
+        spec.output("out")
+
+    def compute(self, op_input, op_output, context):
+        """Add high-frequency modulation to the input signal."""
+        # Receive the input tensor from the source operator
+        input_tensor = op_input.receive("in").get("wave")
+        samples = input_tensor.shape[0]
+
+        # Parameters for high-frequency modulation/noise
+        frequency = 300.0  # Modulation frequency in Hz
+        amplitude = 0.05  # Amplitude of the modulation
+        duration = 1.0  # Duration of signal in seconds
+        sample_time = duration / samples
+        omega = 2 * np.pi * frequency  # Angular frequency
+
+        # Add modulation to the input signal
+        t = np.arange(samples) * sample_time
+        modulation = amplitude * np.sin(omega * t).reshape(samples, 1)
+        modulated_signal = input_tensor + modulation
+
+        # Ensure output is float32
+        modulated_signal = modulated_signal.astype(np.float32)
```

Comment:
add flag to avoid a copy if it is not needed
```suggestion
        modulated_signal = modulated_signal.astype(np.float32, copy=False)
```